Add support for `IterableDataset` in DPO Trainer

h-tonywu opened the pull request:
# What does this PR do?

Add support for `IterableDataset` in DPO Trainer.

## Notes

Waiting for the original issue to be discussed before moving on with the (eventual) tests.

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

qgallouedec commented:
Like this one: https://github.com/huggingface/trl/blob/64b27eadbbe16c7a3a0255a9c795596691841756/tests/test_sft_trainer.py#L1241

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_3559). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

h-tonywu commented:
Thanks for the super quick answer @qgallouedec! 🤗

I have fixed a typing issue and added the requested unit test. Can you run the CI tests once more please?
